clients/grailfiltersegments/filtersegments.go

```diff
@@ -0,0 +1,228 @@
+	}
+
+	return result, nil
+
+}
+
+func (c Client) Upsert(ctx context.Context, uid string, data []byte) (Response, error) {
+	existingResourceResp, err := c.client.Get(ctx, uid, rest.RequestOptions{})
+	if err != nil {
+		return Response{}, fmt.Errorf("failed to get filter segments resource with id %s: %w", uid, err)
+	}
+
+	if existingResourceResp.StatusCode == http.StatusNotFound {
+		newResourceResp, err := c.client.Create(ctx, data)
+		if err != nil {
+			return Response{}, fmt.Errorf("failed to create filter segments resource: %w", err)
+		}
+
+		newResourceBody, err := io.ReadAll(newResourceResp.Body)
+		if err != nil {
+			logr.FromContextOrDiscard(ctx).Error(err, bodyReadErrMsg)
+			return Response{}, api.NewAPIErrorFromResponseAndBody(newResourceResp, newResourceBody)
+		}
+
+		return api.NewResponseFromHTTPResponseAndBody(newResourceResp, newResourceBody), nil
+	}
+
+	existingResourceBody, err := io.ReadAll(existingResourceResp.Body)
+	if err != nil {
+		logr.FromContextOrDiscard(ctx).Error(err, bodyReadErrMsg)
+		return Response{}, api.NewAPIErrorFromResponseAndBody(existingResourceResp, existingResourceBody)
+	}
+
+	type respWithVersion struct {
+		Version int `json:"version"`
+	}
+
+	var currentVersion respWithVersion
+	err = json.Unmarshal(existingResourceBody, &currentVersion)
+	if err != nil {
+		return Response{}, fmt.Errorf("unable to unmarshal data: %w", err)
+	}
+	if currentVersion.Version == 0 {
+		return Response{}, fmt.Errorf("missing version field in API response")
+	}
+
+	updateResourceResp, err := c.client.Update(ctx, uid, data, rest.RequestOptions{QueryParams: map[string][]string{
+		"optimistic-locking-version": {fmt.Sprint(currentVersion.Version)},
+	}})
+
+	if err != nil {
+		return Response{}, fmt.Errorf("failed to update filter segments resource with id %s and version %d: %w", uid, currentVersion.Version, err)
+	}
+
+	updateResourceBody, err := io.ReadAll(updateResourceResp.Body)
+	if err != nil {
+		logr.FromContextOrDiscard(ctx).Error(err, bodyReadErrMsg)
+		return Response{}, api.NewAPIErrorFromResponseAndBody(updateResourceResp, updateResourceBody)
+	}
+
+	return api.NewResponseFromHTTPResponseAndBody(updateResourceResp, updateResourceBody), nil
+}
+
+func (c Client) Delete(ctx context.Context, id string) (Response, error) {
+	if id == "" {
+		return Response{}, errors.New("missing required id")
+	}
+	resp, err := c.client.Delete(ctx, id)
+	if err != nil {
+		return Response{}, fmt.Errorf("failed to get filtersegment resource with id %s: %w", id, err)
+	}
+	if resp != nil && resp.Body != nil {
+		defer resp.Body.Close()
+	}
+
+	if !rest.IsSuccess(resp) {
+		return Response{}, api.NewAPIErrorFromResponse(resp)
+	}
+	return api.NewResponseFromHTTPResponse(resp), nil
+}
+
+var getRequestOptions = rest.RequestOptions{
+	CustomShouldRetryFunc: rest.RetryIfTooManyRequests,
+	QueryParams: map[string][]string{
+		"add-fields": {"INCLUDES", "VARIABLES", "RESOURCECONTEXT"},
```

Comment:
These differ from the other fields, why is that? Is it not necessary?

Reply:
By default, not all parts of the configuration are included. If a user doesn't specify what exactly it needs, we assume that it requests everything.

Reply:
Moved to the raw client